central/globaldb/v2backuprestore/restore/postgres.go

```diff
@@ -0,0 +1,237 @@
+}
+
+// LoadRestore a Postgres database from a dump
+func LoadRestore(dumpPath string) error {
+	log.Info("Starting Postgres Restore")
+	sourceMap, config, err := globaldb.GetPostgresConfig()
+	if err != nil {
+		log.Fatalf("Could not parse postgres config: %v", err)
+		return err
+	}
+
+	// Now recreate the DB
+	err = createDB(sourceMap, config)
+	if err != nil {
+		log.Fatalf("Could not create restore database: %v", err)
+		return err
+	}
+
+	// Execute the restore on the temporary restore database
+	err = runRestore(dumpPath, sourceMap, config)
+	if err != nil {
+		log.Errorf("Could not load the Postgres backup: %v", err)
+		return err
+	}
+
+	log.Info("Postgres Restore Complete")
+	return nil
+}
+
+func renameRestoreDB(connectPool *pgxpool.Pool, updatedDB, primaryDB string) error {
+	log.Info("Flipping the restored DB to the active DB")
+	sqlStmt := fmt.Sprintf("ALTER DATABASE %s RENAME TO %s", updatedDB, primaryDB)
+
+	_, err := connectPool.Exec(context.TODO(), sqlStmt)
+
+	return err
+}
+
+// SwitchToRestoredDB - switches the restore DB to be the active DB
+func SwitchToRestoredDB() error {
+	log.Info("Switching to restored database")
+	sourceMap, config, err := globaldb.GetPostgresConfig()
+	if err != nil {
+		log.Fatalf("Could not parse postgres config: %v", err)
+		return err
+	}
+
+	// connect on template1
+	connectPool := adminPool(config)
+
+	// Restore succeeded to the separate DB, so we need to drop the original in order to rename
+	// the new one.
+	err = dropDB(sourceMap, config, config.ConnConfig.Database)
+	if err != nil {
+		log.Errorf("Could not drop the DB: %v", err)
+		return err
+	}
+
+	// rename central_restore to postgres
+	err = renameRestoreDB(connectPool, restoreDB, config.ConnConfig.Database)
+	if err != nil {
+		log.Errorf("Could not rename the DB: %v", err)
+		return err
+	}
+
+	// Close the connection pool on template1
+	connectPool.Close()
+
+	return nil
+}
+
+func adminPool(config *pgxpool.Config) *pgxpool.Pool {
+	var err error
+	var postgresDB *pgxpool.Pool
+
+	// Clone config to connect to template DB
+	tempConfig := config.Copy()
+
+	// Need to connect on a static DB so we can rename the used DBs.
+	tempConfig.ConnConfig.Database = connectDB
+
+	if err := retry.WithRetry(func() error {
+		postgresDB, err = pgxpool.ConnectConfig(context.Background(), tempConfig)
+		return err
+	}, retry.Tries(postgresOpenRetries), retry.BetweenAttempts(func(attempt int) {
+		time.Sleep(postgresTimeBetweenRetries)
+	}), retry.OnFailedAttempts(func(err error) {
+		log.Errorf("open database: %v", err)
+	})); err != nil {
+		log.Fatalf("Timed out trying to open database: %v", err)
+	}
+
+	return postgresDB
+}
+
+// CheckIfRestoreDBExists - checks to see if a restore database exists
+func CheckIfRestoreDBExists() bool {
+	log.Info("CheckIfRestoreDBExists")
+	_, config, err := globaldb.GetPostgresConfig()
+	if err != nil {
+		log.Fatalf("Could not parse postgres config: %v", err)
+		return false
+	}
+	// connect on template1
```

Comment:
what is `template1`?

Reply:
template1 is a database that automatically gets started with postgres.  Since I need to rename the restore database to the postgres database name we are using, I had to connect to something.  I did write [ROX-11272](https://issues.redhat.com/browse/ROX-11272) to reserve the postgres database for admin type connections such as these and use a different database for general processing.   I will capture and clarify that with a comment and a TODO.